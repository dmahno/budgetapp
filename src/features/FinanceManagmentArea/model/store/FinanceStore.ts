// FinanceStore.ts
import {makeAutoObservable} from 'mobx';
import {message} from 'antd';

import {getDatabase} from '../lib/getDatabase/getDatabase';

interface IOperation {
  id?: number;
  amount: number;
  description: string;
  type: 'add' | 'exclude';
  date: Date;
}

export class FinanceStore {
  operationsList: IOperation[] = [];

  constructor() {
    makeAutoObservable(this);
    this.loadOperations();
  }

  async loadOperations() {
    const database = await getDatabase();
    this.operationsList = await database.getAll('operations');
  }

  async addOperation(
    amount: number,
    description: string,
    type: 'add' | 'exclude',
  ) {
    if (type === 'exclude' && amount > this.totalSum) {
      message.error(
        `Недостаточно средств. Максимальная сумма для списания: ${this.totalSum} ₽`,
      );
      return;
    }

    const database = await getDatabase();
    const newOperation: IOperation = {
      amount,
      description,
      type,
      date: new Date(),
    };

    const operationId = (await database.add(
      'operations',
      newOperation,
    )) as number;
    this.operationsList.push({...newOperation, id: operationId});
  }

  async deleteOperation(operationId: number) {
    const database = await getDatabase();
    await database.delete('operations', operationId);
    this.operationsList = this.operationsList.filter(
      (operation) => operation.id !== operationId,
    );
  }

  get totalSum() {
    return this.operationsList.reduce((accumulatedSum, operation) => {
      return operation.type === 'add'
        ? accumulatedSum + operation.amount
        : accumulatedSum - operation.amount;
    }, 0);
  }
}

export const financeStore = new FinanceStore();
