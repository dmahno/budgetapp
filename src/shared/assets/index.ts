export type TImageKeys = '1.jpg' | '2.jpg' | '3.jpg' | '4.jpg' | '5.jpg';

import image1 from './images/1.jpg';
import image2 from './images/2.jpg';
import image3 from './images/3.jpg';
import image4 from './images/4.jpg';
import image5 from './images/5.jpg';

export const images: Record<TImageKeys, string> = {
  '1.jpg': image1,
  '2.jpg': image2,
  '3.jpg': image3,
  '4.jpg': image4,
  '5.jpg': image5,
};
