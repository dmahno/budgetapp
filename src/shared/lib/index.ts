export {formatNumber} from './formatNumber/formatNumber';
export {parseNumber} from './parseNumber/parseNumber';
