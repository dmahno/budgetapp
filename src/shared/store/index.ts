export * from './AuthStore/AuthStore';
export * from './ThemeStore/ThemeStore';
