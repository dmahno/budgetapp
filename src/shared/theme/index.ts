export {configDarkTheme} from './tokens/configDarkTheme';
export {configLightTheme} from './tokens/configLightTheme';
