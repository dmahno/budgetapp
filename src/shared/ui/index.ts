export {Logo} from './Logo/Logo';
export {ReactRouterCustomLink} from './ReactRouterCustomLink/ReactRouterCustomLink';
export {DynamicIcon} from './DynamicIcon/DynamicIcon';
export {TapButton} from './TapButton/TapButton';
