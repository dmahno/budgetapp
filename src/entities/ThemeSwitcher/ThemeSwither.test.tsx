// ThemeSwitcher.test.tsx
import {render, screen, fireEvent} from '@testing-library/react';
import {vi} from 'vitest';

import {themeStore} from 'shared/store';

import {ThemeSwitcher} from './ThemeSwitcher';

vi.mock('shared/store', () => ({
  themeStore: {
    isDark: false,
    toggleTheme: vi.fn(),
  },
}));

describe('ThemeSwitcher Component', () => {
  beforeEach(() => {
    vi.resetAllMocks();
  });

  it('renders Light icon when isDark is false', () => {
    themeStore.isDark = false;
    render(<ThemeSwitcher />);

    const lightIcon = screen.getByTestId('light-icon');
    expect(lightIcon).toBeInTheDocument();

    const darkIcon = screen.queryByTestId('dark-icon');
    expect(darkIcon).not.toBeInTheDocument();
  });

  it('renders Dark icon when isDark is true', () => {
    themeStore.isDark = true;
    render(<ThemeSwitcher />);

    const darkIcon = screen.getByTestId('dark-icon');
    expect(darkIcon).toBeInTheDocument();

    const lightIcon = screen.queryByTestId('light-icon');
    expect(lightIcon).not.toBeInTheDocument();
  });

  it('calls toggleTheme when button is clicked', () => {
    themeStore.isDark = false;

    render(<ThemeSwitcher />);
    const button = screen.getByRole('button');
    fireEvent.click(button);

    expect(themeStore.toggleTheme).toHaveBeenCalledTimes(1);
  });

  it('matches the snapshot', () => {
    const {asFragment} = render(<ThemeSwitcher />);
    expect(asFragment()).toMatchSnapshot();
  });
});
