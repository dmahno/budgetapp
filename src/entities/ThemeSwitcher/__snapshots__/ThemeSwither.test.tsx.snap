// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ThemeSwitcher Component > matches the snapshot 1`] = `
<DocumentFragment>
  <button
    class="_button_0302f3 _size-s_0302f3 _appearance-thirdly_0302f3"
  >
    <span
      class="_icon_0302f3"
    >
      <svg
        data-testid="light-icon"
        fill="none"
        height="24"
        viewBox="0 0 24 24"
        width="24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          clip-rule="evenodd"
          d="M13 2C13 1.44772 12.5523 1 12 1C11.4477 1 11 1.44772 11 2V3C11 3.55228 11.4477 4 12 4C12.5523 4 13 3.55228 13 3V2ZM13 21C13 20.4477 12.5523 20 12 20C11.4477 20 11 20.4477 11 21V22C11 22.5523 11.4477 23 12 23C12.5523 23 13 22.5523 13 22V21ZM2 11C1.44772 11 1 11.4477 1 12C1 12.5523 1.44772 13 2 13H3C3.55228 13 4 12.5523 4 12C4 11.4477 3.55228 11 3 11H2ZM21 11C20.4477 11 20 11.4477 20 12C20 12.5523 20.4477 13 21 13H22C22.5523 13 23 12.5523 23 12C23 11.4477 22.5523 11 22 11H21ZM8 12C8 9.79086 9.79086 8 12 8C14.2091 8 16 9.79086 16 12C16 14.2091 14.2091 16 12 16C9.79086 16 8 14.2091 8 12ZM12 6C8.68629 6 6 8.68629 6 12C6 15.3137 8.68629 18 12 18C15.3137 18 18 15.3137 18 12C18 8.68629 15.3137 6 12 6ZM19.7782 5.63597C20.1687 5.24544 20.1687 4.61228 19.7782 4.22176C19.3876 3.83123 18.7545 3.83123 18.3639 4.22176L17.6568 4.92886C17.2663 5.31939 17.2663 5.95255 17.6568 6.34308C18.0474 6.7336 18.6805 6.7336 19.071 6.34308L19.7782 5.63597ZM6.34312 19.071C6.73365 18.6805 6.73365 18.0473 6.34312 17.6568C5.9526 17.2663 5.31943 17.2663 4.92891 17.6568L4.2218 18.3639C3.83128 18.7544 3.83128 19.3876 4.2218 19.7781C4.61233 20.1686 5.24549 20.1686 5.63602 19.7781L6.34312 19.071ZM19.7782 19.7784C19.3877 20.1689 18.7546 20.1689 18.364 19.7784L17.6569 19.0713C17.2664 18.6808 17.2664 18.0476 17.6569 17.6571C18.0474 17.2666 18.6806 17.2666 19.0711 17.6571L19.7782 18.3642C20.1688 18.7547 20.1688 19.3879 19.7782 19.7784ZM6.34322 6.34337C5.95269 6.73389 5.31953 6.73389 4.929 6.34337L4.2219 5.63626C3.83137 5.24574 3.83137 4.61257 4.2219 4.22205C4.61242 3.83152 5.24559 3.83152 5.63611 4.22205L6.34322 4.92915C6.73374 5.31968 6.73374 5.95284 6.34322 6.34337Z"
          fill-rule="evenodd"
        />
      </svg>
    </span>
  </button>
</DocumentFragment>
`;
